prover/src/lib.rs

```diff
@@ -192,44 +194,72 @@ pub trait Prover {
+        // 0 ----- instantiate AIR and prover channel ---------------------------------------------
 
-    /// Performs the actual proof generation procedure, generating the proof that the provided
-    /// execution `trace` is valid against this prover's AIR.
-    /// TODO: make this function un-callable externally?
-    #[doc(hidden)]
-    fn generate_proof<E>(&self, mut trace: Self::Trace) -> Result<StarkProof, ProverError>
+        // serialize public inputs; these will be included in the seed for the public coin
+        let pub_inputs = self.get_pub_inputs(&trace);
+        let pub_inputs_elements = pub_inputs.to_elements();
+
+        // create an instance of AIR for the provided parameters. this takes a generic description
+        // of the computation (provided via AIR type), and creates a description of a specific
+        // execution of the computation for the provided public inputs.
+        let air = Self::Air::new(trace.info().clone(), pub_inputs, self.options().clone());
+
+        // create a channel which is used to simulate interaction between the prover and the
+        // verifier; the channel will be used to commit to values and to draw randomness that
+        // should come from the verifier.
+        let mut channel =
+            ProverChannel::<Self::Air, Self::BaseField, Self::HashFn, Self::RandomCoin>::new(
+                &air,
+                pub_inputs_elements,
+            );
+
+        // 1 ----- Commit to the execution trace --------------------------------------------------
+
+        // build computation domain; this is used later for polynomial evaluations
+        let lde_domain_size = air.lde_domain_size();
+        let trace_length = air.trace_length();
+        let domain = info_span!("build_domain", trace_length, lde_domain_size)
+            .in_scope(|| StarkDomain::new(&air));
+        assert_eq!(domain.lde_domain_size(), lde_domain_size);
+        assert_eq!(domain.trace_length(), trace_length);
+
+        // commit to the main trace segment
+        let (trace_lde, trace_polys) =
+            self.build_main_trace_and_commit(&trace, &domain, &mut channel);
+
+        self.generate_proof(trace_lde, trace_polys, domain, &air, None, channel)
+    }
+
+    /// Returns a STARK proof attesting to a correct execution of a computation defined by the
+    /// provided trace. The trace is expected to contain an auxiliary trace segment.
+    ///
+    /// The returned [StarkProof] attests that the specified `trace` is a valid execution trace of
+    /// the computation described by [Self::Air](Prover::Air) and generated using some set of
+    /// secret and public inputs. Public inputs must match the value returned from
+    /// [Self::get_pub_inputs()](Prover::get_pub_inputs) for the provided trace.
+    fn prove_with_aux_trace<E, ATB>(
+        &self,
+        trace: Self::Trace,
+        mut aux_trace_builder: ATB,
+    ) -> Result<(StarkProof, Option<AuxProof<ATB>>), ProverError>
```

Comment:
A couple of thoughts here:

1. I think we should try to combine the `prove()` and `prove_with_aux_trace()` methods so that we have a single `prove()` method.
2. I think we should return a single proof object from this method (rather than a main proof and an optional auxiliary proof).

For the second point, we could make sure that `AuxProof` is serializable and deserializable and then just make it a part of the main proof (we could rename `StarkProof` into just `Proof` to make the naming consistent with the content).

For the first point, I'm not sure yet what the best solution would be. One potential approach is to provide `AuxTraceBuilder` as an optional parameter. The method then would look something like this:
```Rust
fn prove<E, ATB>(
    &self,
    trace: Self::Trace,
    aux_trace_builder: ATB,
) -> Result<Proof, ProverError>
where
    E: FieldElement<BaseField = Self::BaseField>,
    ATB: AuxTraceBuilder<AuxRandElements<E> = Self::AuxRandElements<E>>,
    Self::Air: Air<AuxRandElements<E> = AuxRandElements<ATB, E>>,
```
I think this could work, but there are two downsides to this:
1. We need to make `prove()` generic over `E` and thus, when deserializing `Proof` object, we need to know which extension filed was used to generate the proof. This is not a big deal (we handle hash function options like this already) - but it does add a bit more work.
2. I don't love making the function generic over `AxTraceBuilder` (and to some extent needing to pass a new `aux_trace_builder` on every call to `prove()`). A potential way around this is to make `AuxTraceBuilder` an associated type and then add a constructor method to the `Prove` trait to build it similar to how we handle `TraceLde` construction. But I understand that there may be challenges with this as well.

The other potential option is to get rid of `AuxTraceBuilder` trait and go back to the `Trace::build_aux_segment()` function. As you've pointed out, we'd need to add two extra associated types to the `Trace` trait:

```Rust
pub trait Trace: Sized {
    type BaseField: StarkField;
    type AuxRandElements<E: FieldElement<BaseField = Self::BaseField>>;
    type AuxProof: Serializable + Deserializable
```
The complication here could be related to enforcing that `Air::AuxRandElements` and `Trace::AuxRandElements` are the same type.

Reply:
- Merged both `prove_()` functions into a single non-generic `prove()` (as we previously had)
- `StarkProof` now contains a serialized `AuxProof`

The one quirk that I was not able to remove is that provers that prove a trace with no auxiliary segment will need to set `type AuxTraceBuilder<E> = EmptyAuxTraceBuilder`, where `EmptyAuxTraceBuilder` is a dummy type that implements the `AuxTraceBuilder` trait. This is ultimately a consequence of `prove()` not being generic:
- We can't write the `Self::AuxTraceBuilder<E>: AuxTraceBuilder<AuxRandElements<E> = AuxRandElementsProver<Self, E>, AuxProof = Self::AuxProof>` trait bound on `prove()` (since there is no `E` generic type)
- We also can't write it on `generate_proof<E>()`, since then `prove()` can't call it anymore (because `prove()` doesn't have the constraint)
- So we're forced to put that trait bound directly on the `Prover`'s associated type definition. Which means that implementors can't use `type AuxTraceBuilder = ()` anymore, since `()` doesn't implemement the `AuxTraceBuilder` trait.

Also, a consequence of putting the serialized `AuxProof` in `StarkProof` (now renamed `Proof`) is that the verifier needs to know the type that it will deserialize in. Hence, I wasn't able to avoid having 2 `verify()` methods (one that provides an `AuxTraceVerifier` which defines the `AuxProof` type, and on that doesn't).